myhpom/tasks.py

```diff
@@ -0,0 +1,119 @@
+import json
+import requests
+from django.core.mail import send_mail
+from django.conf import settings
+from django.template.loader import get_template
+from django.template import Context
+from celery import shared_task
+from celery.task import Task
+from celery.signals import task_failure
+from myhpom.models import AdvanceDirective, DocumentUrl, CloudFactoryRun, CloudFactoryUnit
+
+
+@shared_task
+class CloudFactorySubmitAdvanceDirectiveRun(Task):
+    def run(self, ad_id, line_id, document_host, callback_url):
+        """
+        ## Arguments
+        * ad_id = the database id for the AdvanceDirective
+        * line_id = the id of the CloudFactory production line to submit this request to.
+        * document_host = the host (incl. protocol/scheme) to use for the DocumentURL
+        * callback_url = the URL to use with the callback request
+
+        ## Task Process
+        - create an expiring DocumentUrl for the AdvanceDirective
+        - submit the DocumentUrl to the CloudFactory (test) endpoint
+        - store the status of the submitted document
+        - if the AdvanceDirective no longer exists, abort the task without an Exception.
+            - the AdvanceDirective might no longer exist by the time celery picks it up,
+            - or the document it points to might have been removed.
+        - if an error occurs in this process, send support email.
+        """
+        cf_run = CloudFactoryRun.objects.create(line_id=line_id, callback_url=callback_url or "")
+
+        # If the AdvanceDirective has been deleted, abort the task with a message
+        # -- no need to admin support email, this is an expected case.
+        try:
+            ad = AdvanceDirective.objects.get(id=ad_id)
+        except:
+            cf_run.status = 'ABORTED'
```

Comment:
We could use cf_run.status only to keep status responses from CloudFactory. At present, I think the choices are "Processing" and "Processed". (When you abort, the run status is still "Processing" because it does it async, and when the abort has finished it's "Processed". The Unit.status is "Aborting" after you post the abort, and "Aborted" when they have finished aborting.)

I think you're right that we should have a way to save our own internal status (such as this case, when the AD has been deleted), and maybe we shouldn't pack that into the field that holds the CloudFactory status. But that suggests another field. Unless we just preemptively delete the run object when the AD has been deleted.

Reply:
It kind of does suggest a different field, or breaking out the two uses: a way to get CF's results (a method that returns the payload that we review directly, or another field that we store some final state (success/fail) if we actually want to write queries against it), and one to store our sense of the request status (initialized/queued, aborted, errored, succeeded).

This later field is important b/c in our future ticket (254 I guess I don't another that makes sense) we need to ensure that we recover gracefully when a run fails.